### ISSUE TITLE
rfc14: late binding for task and/or command?

### ISSUE THREAD

**Issue by trws** [flux-framework/rfc#200]
An interesting point came up in a discussion on machines with heterogeneous nodes this morning.  Currently, we have early or late binding of task counts based on whether they are using a per-slot or per-resource count.  The idea was to have late binding of the task or command executed based on the type of resource received after scheduling.  I'm not sure exactly how we would spell this, but it could be something like two slot types where we allow one *or* the other, and then the task that references the winning slot type gets run?

The specific issue being solved is machines like Trinity, where you have two fundamentally different architectures between the nodes, where you might be able to run one common binary but only by building to a severely limited common denominator, so being able to say "I have two versions, run the one that makes sense" could be quite a boon.  Moreso perhaps on machines where some nodes have different GPU types, or even CPU instruction sets than others.  Thoughts?

**Comment by grondo** [comment]
Interesting idea! How the reader of the final jobspec+R (the job shell), determine the task/command spec that makes sense when it is time to execute the job. Making it very explicit would be helpful :wink:

Are you thinking of a labeled list of alternates, and the scheduler would emit the label along with R to indicate which alternate was matched?

This seems similar to submitting two separate jobs but telling the system to only run the first job that is scheduled. (I'm not sure that silly observation is useful, but perhaps evidence of another use case?)

**Comment by trws** [comment]
Essentially yes, I was thinking in terms of the slot labels since that's a mechanism we already have, with the addition of something to say "either this slot type or that one."  That way we could use the label system we have for slots and get to the same place.  Alternately we could have a way to put a "match" clause on the task to choose the best task or command to run maybe?

I think you're right that it's essentially the same as putting in two jobs where one supersedes the other if it gets scheduled.  The main difference I think would be in expressibility.  This way you could say something like "I need one node, don't care what kind, but if it's this kind run this command rather than the default one" or perhaps "I can run on 3 haswell nodes, but need 34 KNL nodes to do the same job."  I think it comes down to something like a generalization of the count/range stuff we did for resources, but applying to the actual job specification as well as the resource component.

**Comment by grondo** [comment]
Yes, I like it. It also gives the user a single jobid on whic to call "wait", so has a huge benefit over separate exclusive-or jobs.

I'm not sure we've defined  if/how slot labels are propagated by a scheduler into R, so that is something we might have to explore.

**Comment by trws** [comment]
I don’t see why we couldn’t do that necessarily. It will need a good think on how the APIs get tweaked, but I like the chances of being able to do this after the discussion.

Sent from my iPhone

On Aug 28, 2019, at 5:02 AM, Dong H. Ahn <notifications@github.com<mailto:notifications@github.com>> wrote:


From the point of view of our graph based scheduler, At first glance, the scheduler will need to split the resource section per each equivalent (or pararallel for lack of good names) slot type in preference order and perform a match one by one until a match.

The matched resources are emitted into R which needs to be extended to encode slots and slot labels.

Essentially this is two virtual jobspecs that are merged into a single spec.

So grom the implementation point a view, I think the scheduler will greatly appreciate if a jobspec library does this split for the scheduler and let it iterate.
